src/Workspaces/Core/Portable/Serialization/SerializableSourceText.cs

```diff
@@ -49,21 +57,45 @@ private SerializableSourceText(ITemporaryTextStorageWithName? storage, SourceTex
         {
             Debug.Assert(storage is null != text is null);
 
-            Storage = storage;
-            Text = text;
+            _storage = storage;
+            _text = text;
         }
 
-        public ImmutableArray<byte> GetChecksum()
+        /// <summary>
+        /// Returns the strongly referenced SourceText if we have it, or tries to retrieve it from the weak reference if
+        /// it's still being held there.
+        /// </summary>
+        /// <returns></returns>
+        private SourceText? TryGetText()
+            => _text ?? _computedText.GetTarget();
+
+        public ImmutableArray<byte> GetChecksum(CancellationToken cancellationToken)
         {
-            return Text?.GetChecksum() ?? Storage!.GetChecksum();
+            return TryGetText()?.GetChecksum() ?? _storage!.GetChecksum();
         }
 
         public async ValueTask<SourceText> GetTextAsync(CancellationToken cancellationToken)
         {
-            if (Text is not null)
-                return Text;
+            var text = TryGetText();
+            if (text != null)
+                return text;
 
-            return await Storage!.ReadTextAsync(cancellationToken).ConfigureAwait(false);
+            // Read and cache the text from the storage object so that other requests may see it if still kept alive by something.
+            text = await _storage!.ReadTextAsync(cancellationToken).ConfigureAwait(false);
+            _computedText.SetTarget(text);
+            return text;
+        }
+
+        public SourceText GetText(CancellationToken cancellationToken)
+        {
+            var text = TryGetText();
+            if (text != null)
+                return text;
+
+            // Read and cache the text from the storage object so that other requests may see it if still kept alive by something.
```

Comment:
So at this layer of the SerializeableSourceText directly, this could read the same thing more than once and end up with different values? But we're letting the serializer service at a higher layer deal with the dedup? In no way questioning the design, just mentally confirming the layering.

Reply:
yes.  the idea is that the higher layer keeps the mapping from Checksum->SerializableSourceText (so multiple things asking for that same checksum will always get teh same SerializableSourceText  instance).  Then, here, what we're at least trying to do is say that if one other concurrent call already read this value, it can be seen by other concurrent calls.

but it's still possible to get no win here if teh concurrent calls came in at the exact same time into  this and requested the SOurceText.  We could avoid that with a lock though.